include/nbl/system/ISystem.h

```diff
@@ -0,0 +1,244 @@
+            request->set_skip();
+        }
+
+        bool ready() const { return !request || request->ready; }
+        bool valid() const { return valid_flag; }
+
+        void wait()
+        {
+            asyncQ->waitForRequestCompletion(request[0]);
+        }
+    };
+
+    template <typename T>
+    class SFuture : public SFutureBase
+    {
+        alignas(T) uint8_t storage[sizeof(T)];
+
+    public:
+        ~SFuture()
+        {
+            if (valid_flag)
+                getStorage()->~T();
+        }
+
+        T* getStorage() { return reinterpret_cast<T*>(storage); }
+        T get()
+        {
+            wait();
+            assert(valid_flag);
+            valid_flag = false;
+            T* ptr = getStorage();
+            return std::move(ptr[0]);
+        }
+    };
+
+    // @sadiuk add more request types if needed
+    enum E_REQUEST_TYPE
+    {
+        ERT_CREATE_READ,
+        ERT_CREATE_WRITE,
+        ERT_READ,
+        ERT_WRITE
+    };
+    template <E_REQUEST_TYPE RT>
+    struct SRequestParamsBase
+    {
+        static inline constexpr E_REQUEST_TYPE type = RT;
+        SFutureBase* future;
+        void* retval_storage;
+    };
+    struct SRequestParams_CREATE_READ : SRequestParamsBase<ERT_CREATE_READ>
+    {
+        std::filesystem::path filename;
+    };
+    struct SRequestParams_CREATE_WRITE : SRequestParamsBase<ERT_CREATE_WRITE>
+    {
+        std::filesystem::path filename;
+    };
+    struct SRequestParams_READ : SRequestParamsBase<ERT_READ>
+    {
+        IReadFile* file;
+        void* buffer;
+        size_t offset;
+        size_t size;
+    };
+    struct SRequestParams_WRITE : SRequestParamsBase<ERT_WRITE>
+    {
+        IWriteFile* file;
+        const void* buffer;
+        size_t offset;
+        size_t size;
+    };
+    struct SRequestType : SSkippableRequestBase
+    {
+        SFutureBase* future = nullptr;
+        void* retval_storage = nullptr;
+        E_REQUEST_TYPE type;
+        std::variant<
+            SRequestParams_CREATE_READ,
+            SRequestParams_CREATE_WRITE,
+            SRequestParams_READ,
+            SRequestParams_WRITE
+        > params;
+
+        void set_skip()
+        {
+            write_lock_guard<> lk(rwlock);
+            set_skip_no_lock();
+            future->request = nullptr;
+        }
+    };
+
+    class CAsyncQueue : public ISkippableAsyncQueueDispatcher<CAsyncQueue, SRequestType, 256u>
+    {
+        using base_t = ISkippableAsyncQueueDispatcher<CAsyncQueue, SRequestType, 256u>;
+
+    public:
+        explicit CAsyncQueue(ISystem* owner, core::smart_refctd_ptr<ISystemCaller>&& caller) : m_owner(owner), m_caller(std::move(caller)) {}
+
+        template <typename RequestParams>
+        void request_impl(SRequestType& req, RequestParams&& params)
+        {
+            req.type = params.type;
+            req.params = std::move(params);
+            req.future = params.future;
+            req.retval_storage = params.retval_storage;
```

Comment:
you can get the storage pointer from the future itself... why store it separately?

Reply:
also... you have both `req.params.future` and `req.future` together.

Reply:
> you can get the storage pointer from the future itself... why store it separately?

I have pointer to SFutureBase, not to SFuture. SFuture is a template so cant really have a pointer to it

Reply:
but why does your system need to know about the pointer to `retval_storage` if you had a `notify` function?
https://github.com/Crisspl/Nabla/pull/4#discussion_r622660956

Reply:
originally to create returned object on it, but with `notify(args...)` and being able to store pointer to `SFuture<whatever_type_i_want>` in request params i dont need it any more